src/software/ai/hl/stp/tactic/chip_tactic_test.cpp

```diff
@@ -2,15 +2,292 @@
+
+    Robot robot = Robot(0, Point(0, 0), Vector(0, 0), Angle::zero(),
+                        AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    ChipTactic tactic = ChipTactic(world.ball(), true);
+
+    tactic.updateRobot(robot);
+    tactic.updateControlParams(Point(3, 1), Point(0, 1), 1.0);
+    auto action_ptr = tactic.getNextAction();
+
+    // Check an action was returned (the pointer is not null)
+    EXPECT_TRUE(action_ptr);
+
+    auto chip_action = std::dynamic_pointer_cast<ChipAction>(action_ptr);
+    ASSERT_NE(chip_action, nullptr);
+    EXPECT_EQ(Point(3, 1), chip_action->getChipOrigin());
+}
+
+TEST(ChipTacticTest, getChipDistanceMeters)
+{
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+    world =
+        ::Test::TestUtil::setBallPosition(world, Point(0, 0), Timestamp::fromSeconds(0));
+
+    Robot robot = Robot(0, Point(0, 0), Vector(0, 0), Angle::zero(),
+                        AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    ChipTactic tactic = ChipTactic(world.ball(), true);
+
+    tactic.updateRobot(robot);
+    tactic.updateControlParams(Point(3, 1), Point(0, 1), 2.0);
+    auto action_ptr = tactic.getNextAction();
+
+    // Check an action was returned (the pointer is not null)
+    EXPECT_TRUE(action_ptr);
+
+    auto chip_action = std::dynamic_pointer_cast<ChipAction>(action_ptr);
+    ASSERT_NE(chip_action, nullptr);
+    EXPECT_EQ(2.0, chip_action->getChipDistanceMeters());
+}
+
+TEST(ChipTacticTest, robot_behind_ball_chipping_towards_positive_x_positive_y)
+{
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+    world =
+        ::Test::TestUtil::setBallPosition(world, Point(0, 0), Timestamp::fromSeconds(0));
+
+    Robot robot = Robot(0, Point(-0.3, 0), Vector(0, 0), Angle::zero(),
+                        AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    ChipTactic tactic = ChipTactic(world.ball(), true);
+
+    Ball ball({0, 0}, robot.position().toVector(), Timestamp::fromSeconds(0));
+
+    tactic.updateRobot(robot);
+    tactic.updateControlParams(Point(0, 0), Point(1, 1), 2.0);
+    auto action_ptr = tactic.getNextAction();
+
+    // Check an action was returned (the pointer is not null)
+    ASSERT_TRUE(action_ptr);
+
+    auto chip_action = std::dynamic_pointer_cast<ChipAction>(action_ptr);
+
+    ChipAction expected_chip_action = ChipAction();
+
+    expected_chip_action.updateWorldParams(ball);
+    expected_chip_action.updateControlParams(robot, ball.position(), Angle::fromDegrees(45.0), 2.0);
+
+    // TODO implement compareChipActions
+    // TODO cast action_ptr to ChipAction type
+
+    ASSERT_NE(chip_action, nullptr);
+//    EXPECT_EQ(Point(0, 0), chip_action->getChipOrigin());
+//    EXPECT_EQ(Angle::fromDegrees(45.0), chip_action->getChipDirection());
+//    EXPECT_EQ(2.0, chip_action->getChipDistanceMeters());
+    EXPECT_TRUE(compareChipActions((ChipAction) &action_ptr, &expected_chip_action));
```

Comment:
the `action_ptr` has already been cast to a `ChipAction` above (line 103), so you can just use `chip_action` here.

Reply:
Now using `chip_action`